ghcide/src/Development/IDE/Plugin/TypeLenses.hs

```diff
@@ -340,3 +372,188 @@ pprPatSynTypeWithoutForalls p = pprPatSynType pWithoutTypeVariables
+  { bindings         :: [LocalBinding]
+  , existingSigNames :: [Name]
+  -- ^ Names of existing signatures.
+  -- It is used to hide type lens for existing signatures.
+  --
+  -- NOTE: The location of this name is equal to
+  -- the binding name.
+  --
+  -- Example:
+  -- @
+  -- f :: Int
+  -- f = 42
+  -- @
+  -- The location of signature name `f`(first line) is equal to
+  -- the definition of `f`(second line).
+  }
+
+data GetLocalBindingTypeSigs = GetLocalBindingTypeSigs
+  deriving (Generic, Show, Eq, Ord, Hashable, NFData)
+
+type BindingSigMap = Map.Map Id String
+
+newtype LocalBindingTypeSigsResult = LocalBindingTypeSigsResult ([LocalBindings], BindingSigMap)
+
+instance Show LocalBindingTypeSigsResult where
+  show _ = "<GetTypeResult.local>"
+
+instance NFData LocalBindingTypeSigsResult where
+  rnf = rwhnf
+
+type instance RuleResult GetLocalBindingTypeSigs = LocalBindingTypeSigsResult
+
+localBindingRules :: Recorder (WithPriority Log) -> Rules ()
+localBindingRules recorder = do
+  define (cmapWithPrio LogShake recorder) $ \GetLocalBindingTypeSigs nfp -> do
+    tmr <- use TypeCheck nfp
+    -- we need session here for tidying types
+    hsc <- use GhcSession nfp
+    result <- liftIO $ localBindingType (tmrTypechecked <$> tmr) (hscEnv <$> hsc)
+    pure ([], result)
+
+localBindingType :: Maybe TcGblEnv -> Maybe HscEnv -> IO (Maybe LocalBindingTypeSigsResult)
+localBindingType (Just gblEnv) (Just hsc) = do
+  let locals = findLocalQ (tcg_binds gblEnv)
+      localBindings = mapMaybe findBindingsQ locals
+      bindToSig' = bindToSig hsc (tcg_rdr_env gblEnv)
+      findSigs (LocalBindings bindings _) = fmap findSig bindings
+        where findSig (LocalBinding bindingId _ _) = sequence (bindingId, bindToSig' bindingId)
+  (_, Map.fromList . fromMaybe [] -> sigMap) <-
+    initTcWithGbl hsc gblEnv ghostSpan $ sequence $ concatMap findSigs localBindings
+  pure $ Just (LocalBindingTypeSigsResult (localBindings, sigMap))
+localBindingType _ _                   = pure Nothing
+
+-- | All local bind expression from type checked source.
+findLocalQ :: GenericQ [HsLocalBinds GhcTc]
+findLocalQ = everything (<>) ([] `mkQ` (pure . findWhere) `extQ` findLet)
+  where
+    findWhere :: GRHSs GhcTc (LHsExpr GhcTc) -> HsLocalBinds GhcTc
+    findWhere = grhssLocalBinds
+
+    findLet :: LHsExpr GhcTc -> [HsLocalBinds GhcTc]
+    findLet = findLetExpr . unLoc
+
+    findLetExpr :: HsExpr GhcTc -> [HsLocalBinds GhcTc]
+#if !MIN_VERSION_ghc(9,9,0)
+    findLetExpr (HsLet _ _ binds _ _)       = [binds]
+#else
+    findLetExpr (HsLet _ binds _)           = [binds]
+#endif
+    findLetExpr (HsDo _ _ (unLoc -> stmts)) = concatMap (findLetStmt . unLoc) stmts
+    findLetExpr _                           = []
+
+    findLetStmt :: ExprStmt GhcTc -> [HsLocalBinds GhcTc]
+    findLetStmt (LetStmt _ binds) = [binds]
+    -- TODO(jinser): why `foo <- expr` does not exist
+    -- findLetStmt (BindStmt _ _ expr) = findLetExpr (unLoc expr)
```

Comment:
I'm wondering why this doesn't work, `expr` in `BindStmt` contains Var but not Bind.